oxidation/libparsec/crates/core/src/invite/greeter.rs

```diff
@@ -0,0 +1,669 @@
+            shared_secret_key: self.0.shared_secret_key,
+            cmds: self.0.cmds,
+        }))
+    }
+}
+
+// GreetInProgress2Ctx
+
+#[derive(Debug)]
+struct BaseGreetInProgress2Ctx {
+    token: InvitationToken,
+    claimer_sas: SASCode,
+    shared_secret_key: SecretKey,
+    cmds: AuthenticatedCmds,
+}
+
+impl BaseGreetInProgress2Ctx {
+    fn generate_claimer_sas_choices(&self, size: usize) -> Vec<SASCode> {
+        SASCode::generate_sas_code_candidates(&self.claimer_sas, size)
+    }
+
+    async fn do_signify_trust(&self) -> InviteResult<()> {
+        let rep = self
+            .cmds
+            .send(invite_3b_greeter_signify_trust::Req { token: self.token })
+            .await?;
+
+        match rep {
+            invite_3b_greeter_signify_trust::Rep::AlreadyDeleted => Err(InviteError::AlreadyUsed),
+            invite_3b_greeter_signify_trust::Rep::InvalidState => Err(InviteError::PeerReset),
+            invite_3b_greeter_signify_trust::Rep::NotFound => Err(InviteError::NotFound),
+            invite_3b_greeter_signify_trust::Rep::Ok => Ok(()),
+            invite_3b_greeter_signify_trust::Rep::UnknownStatus { unknown_status, .. } => Err(
+                InviteError::Custom(format!("Backend error during step 3b: {unknown_status}")),
+            ),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct UserGreetInProgress2Ctx(BaseGreetInProgress2Ctx);
+
+impl UserGreetInProgress2Ctx {
+    pub fn claimer_sas(&self) -> &SASCode {
+        &self.0.claimer_sas
+    }
+
+    pub fn generate_claimer_sas_choices(&self, size: usize) -> Vec<SASCode> {
+        self.0.generate_claimer_sas_choices(size)
+    }
+
+    pub async fn do_signify_trust(self) -> InviteResult<UserGreetInProgress3Ctx> {
+        self.0.do_signify_trust().await?;
+
+        Ok(UserGreetInProgress3Ctx(BaseGreetInProgress3Ctx {
+            token: self.0.token,
+            shared_secret_key: self.0.shared_secret_key,
+            cmds: self.0.cmds,
+        }))
+    }
+}
+
+#[derive(Debug)]
+pub struct DeviceGreetInProgress2Ctx(BaseGreetInProgress2Ctx);
+
+impl DeviceGreetInProgress2Ctx {
+    pub fn claimer_sas(&self) -> &SASCode {
+        &self.0.claimer_sas
+    }
+
+    pub fn generate_claimer_sas_choices(&self, size: usize) -> Vec<SASCode> {
+        self.0.generate_claimer_sas_choices(size)
+    }
+
+    pub async fn do_signify_trust(self) -> InviteResult<DeviceGreetInProgress3Ctx> {
+        self.0.do_signify_trust().await?;
+
+        Ok(DeviceGreetInProgress3Ctx(BaseGreetInProgress3Ctx {
+            token: self.0.token,
+            shared_secret_key: self.0.shared_secret_key,
+            cmds: self.0.cmds,
+        }))
+    }
+}
+
+// GreetInProgress3Ctx
+
+#[derive(Debug)]
+struct BaseGreetInProgress3Ctx {
+    token: InvitationToken,
+    shared_secret_key: SecretKey,
+    cmds: AuthenticatedCmds,
+}
+
+impl BaseGreetInProgress3Ctx {
+    async fn do_get_claim_requests(&self) -> InviteResult<Vec<u8>> {
```

Comment:
Take ownership, but that would require returning internal value.
Maybe create a structure that store those value ?

Reply:
Done